### PR TITLE
SDKS-1749 Device Binding QA

### PR DESCRIPTION
**Pull request by spetrov**
# JIRA Ticket

[SDKS-1749](https://bugster.forgerock.org/jira/browse/SDKS-1749) "Device Binding QA"

# Description

- Updated FRUI to support the new `DeviceBindingCallback`, so that we can perform manual testing using the `auth` sample app.
- Updated temporarily the `DeviceBindingCallback` and `DeviceBindingStatus`, so that I don't break the build.  A proper fix for these will be done with [SDKS-2132](https://bugster.forgerock.org/jira/browse/SDKS-2132).
- Added e2e test coverage. Note that these will fail, since the node is not present in our test ID Cloud instance.

### CODE REVIEW

**Comment by jeyanthanperiyasamy** [comment]
LGTM

**Review by jeyanthanperiyasamy — APPROVED** [review]

File: forgerock-auth-ui/src/main/java/org/forgerock/android/auth/ui/callback/DeviceBindingCallbackFragment.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2020 - 2022 ForgeRock. All rights reserved.
```
Comment (witrisna) [inline forgerock-auth-ui/src/main/java/org/forgerock/android/auth/ui/callback/DeviceBindingCallbackFragment.java:2]: Just 2022 should be good.

File: forgerock-auth-ui/src/main/java/org/forgerock/android/auth/ui/callback/DeviceBindingCallbackFragment.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    public View onCreateView(LayoutInflater inflater, ViewGroup container,
+                             Bundle savedInstanceState) {
+        // Inflate the layout for this fragment
+        View view = inflater.inflate(R.layout.fragment_device_binding_callback, container, false);
+        message = view.findViewById(R.id.message);
+        if (node.getCallbacks().size() == 1) { //auto submit if there is one node
+            message.setText(callback.getContent());
+        } else {
+            message.setVisibility(GONE);
+        }
```
Comment (witrisna) [inline forgerock-auth-ui/src/main/java/org/forgerock/android/auth/ui/callback/DeviceBindingCallbackFragment.java:52]: I don't think u need the message and this block of code.

File: forgerock-auth-ui/src/main/java/org/forgerock/android/auth/ui/callback/DeviceBindingCallbackFragment.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+                    if(((DeviceBindingException) e).getStatus() instanceof Timeout) {
+                        callback.setClientError("Timeout");
+                    }
+                    else if(((DeviceBindingException) e).getStatus() instanceof Abort) {
+                        callback.setClientError("Abort");
+                    }
+                    else if(((DeviceBindingException) e).getStatus() instanceof Unsupported) {
+                        callback.setClientError("Unsupported");
+                    }
```
Comment (witrisna) [inline forgerock-auth-ui/src/main/java/org/forgerock/android/auth/ui/callback/DeviceBindingCallbackFragment.java:80]: I believe the SDK will setClientError based on result, so should not neet to set the clientError here.

File: forgerock-auth-ui/src/main/res/layout/fragment_device_binding_callback.xml [hunk 4]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+  ~ Copyright (c) 2019 - 2022 ForgeRock. All rights reserved.
```
Comment (witrisna) [inline forgerock-auth-ui/src/main/res/layout/fragment_device_binding_callback.xml:3]: Just 2022

File: forgerock-auth-ui/src/main/res/layout/fragment_device_binding_callback.xml [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  ~ This software may be modified and distributed under the terms
+  ~ of the MIT license. See the LICENSE file for details.
+  -->
+
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    tools:context=".callback.DeviceProfileCallbackFragment" android:id="@+id/frameLayout">
```
Comment (witrisna) [inline forgerock-auth-ui/src/main/res/layout/fragment_device_binding_callback.xml:14]: Change the context to DeviceBindingCallbackFragment

**Review by witrisna — CHANGES_REQUESTED** [review]

File: forgerock-auth-ui/src/main/java/org/forgerock/android/auth/ui/callback/DeviceBindingCallbackFragment.java (hunk 1 above)
Comment (witrisna) [inline forgerock-auth-ui/src/main/java/org/forgerock/android/auth/ui/callback/DeviceBindingCallbackFragment.java:2]: Just 2022

Comment (spetrov) [reply forgerock-auth-ui/src/main/java/org/forgerock/android/auth/ui/callback/DeviceBindingCallbackFragment.java:2]: 👍🏻 Fixed!

File: forgerock-auth-ui/src/main/java/org/forgerock/android/auth/ui/callback/DeviceBindingCallbackFragment.java (hunk 2 above)
Comment (spetrov) [reply forgerock-auth-ui/src/main/java/org/forgerock/android/auth/ui/callback/DeviceBindingCallbackFragment.java:52]: Right. I removed the message and the block...

File: forgerock-auth-ui/src/main/java/org/forgerock/android/auth/ui/callback/DeviceBindingCallbackFragment.java (hunk 3 above)
Comment (spetrov) [reply forgerock-auth-ui/src/main/java/org/forgerock/android/auth/ui/callback/DeviceBindingCallbackFragment.java:80]: I retested the client error scenarios, and indeed no special logic is needed... so I cleaned this logic... 
Thanks for pointing this out!

File: forgerock-auth-ui/src/main/res/layout/fragment_device_binding_callback.xml (hunk 4 above)
Comment (spetrov) [reply forgerock-auth-ui/src/main/res/layout/fragment_device_binding_callback.xml:3]: 👍🏻

File: forgerock-auth-ui/src/main/res/layout/fragment_device_binding_callback.xml (hunk 5 above)
Comment (spetrov) [reply forgerock-auth-ui/src/main/res/layout/fragment_device_binding_callback.xml:14]: Oops... Wow, eagle eyes you have! 🥇 Fixed that...

File: forgerock-auth-ui/src/main/java/org/forgerock/android/auth/ui/callback/DeviceBindingCallbackFragment.java (hunk 1 above)
Comment (spetrov) [reply forgerock-auth-ui/src/main/java/org/forgerock/android/auth/ui/callback/DeviceBindingCallbackFragment.java:2]: fixed

**Review by witrisna — APPROVED** [review]

**Pull request merged**